CSUB-368 Staking Pallet

CAGS295 opened the pull request:
Description of proposed changes:
This is a trimmed implementation from substrate/frame/staking. This is very likely to be incomplete and broken at some places but, it is a common ground for the next features.

Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

CAGS295 committed:
Staking pallet

CAGS295 committed:
Staking Interface

CAGS295 committed:
staking package

atodorov commented on pallets/staking/Cargo.toml:
```diff
@@ -0,0 +1,40 @@
+    'pallet-staking-substrate/std',
+    'sp-std/std',
+]
+try-runtime = ['frame-support/try-runtime']
```
Not using anything outside workspace for now but please add to `.github/dependabot.yml`

atodorov commented on pallets/staking/src/pallet/impls.rs:
```diff
@@ -0,0 +1,267 @@
+
+		let mut total_imbalance = PositiveImbalanceOf::<T>::zero();
+		// We can now make total validator payout:
+		//TODO
```
Not sure what this TODO is meant to represent. Maybe open a ticket if that will be a future task.

atodorov commented on pallets/staking/src/pallet/mod.rs:
```diff
@@ -0,0 +1,1034 @@
+		/// unless the `origin` falls below _existential deposit_ and gets removed as dust.
+		/// ------------------
+		/// # </weight>
+		#[pallet::call_index(0)]
```
The location of this file itself and also this line will break the pallet discovery script in #883. Will adjust.

atodorov commented on pallets/staking/src/pallet/mod.rs:
```diff
@@ -0,0 +1,1034 @@
+		/// # </weight>
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::bond_extra())]
+		pub fn bond_extra(
```
Note: all of these extrinsics will need tests.

atodorov reviewed (changes requested):
Added initial comments

atodorov commented on pallets/staking/src/pallet/mod.rs:
```diff
@@ -0,0 +1,1034 @@
+
+		// Weight information for extrinsics in this pallet.
+		//TODO finish benchmarks.
+		type WeightInfo: WeightInfo;
```
Can you put these in `pallets/staking/src/weights.rs` and `pallets/staking/src/benchmarking.rs` when you come to it? 

I've updated the parity checking script in #883 to be able to discover the staking pallet and extract extrinsics out of it but it will still look for the benchmarks and weights in fixed locations. See https://github.com/gluwa/creditcoin/actions/runs/3911712880/jobs/6685414925

FTR the script will need to be adjusted a bit more I think.

CAGS295 replied:
I can copy the weights from substrate's staking. What exactly do you need from benchmarking.rs? So far, I can't compile benchmarks because the pallet needs to be 'in use' by the runtime to benchmark it.

atodorov committed:
Temporary exclude staking from the check script until CSUB-390

this pallet is still not in use and benchmarks implementation will
follow afterwards as a separate task.

nathanwhit commented on pallets/staking/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+#[cfg(feature = "std")]
+pub use pallet::GenesisConfig;
+pub use pallet::{Config, Error, Event, Pallet};
+pub use pallet::{
+	__InherentHiddenInstance, __substrate_call_check, __substrate_event_check,
+	__substrate_genesis_config_check, tt_default_parts, tt_error_token,
+};
```
What's the deal w/ these re-exports? I don't see them used anywhere and I get the impression they're internal details of the macro-generated code (that's normally the case for things prefixed w/ `__`)

CAGS295 replied:
It is a matter of preference. `construct_runtime` expects those at the top level. I don't like glob imports, so I expose these only.

nathanwhit replied:
The `construct_runtime` macro accepts the path to the pallet module, the pallet doesn't necessarily need to be defined at the top level. In other words, you can remove these re-exports, and then in the `construct_runtime` call:

```rust
construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = opaque::Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		System: frame_system,
		Timestamp: pallet_timestamp,
		Balances: pallet_balances,
		TransactionPayment: pallet_transaction_payment,
		Sudo: pallet_sudo,
		Creditcoin: pallet_creditcoin,
		Difficulty: pallet_difficulty,
		Rewards: pallet_rewards,
		Scheduler: pallet_scheduler,
		TaskScheduler: pallet_offchain_task_scheduler,
		Staking: pallet_staking::pallet, // path to the actual pallet module
	}
);
```

nathanwhit commented on pallets/staking/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+	__substrate_genesis_config_check, tt_default_parts, tt_error_token,
+};
+
+pub const _STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
```
Why the leading underscore? Also not used anywhere AFAICT

nathanwhit commented on pallets/staking/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+
+use logger;
+
+/// Information regarding the active era (era in used in session).
```
Confused by "era in used in session"

nathanwhit reviewed (commented):
Still reviewing (it's a lot of code) but a couple top-level thoughts:

- Since it seems like a decent chunk of this code is copied from `pallet-staking`, we should be careful of licensing. I think @atodorov knows more about this than I do, but I imagine we need to at least acknowledge that this is adapted and where it's adapted from
- It would be good to know where we differ from `pallet-staking`, to focus review (I still want to review it all, but assuming substrate is well-reviewed, I'd prefer to focus on our changes)
- It feels a _little_ weird that we still depend on `pallet-staking` and that some parts are _almost_ the same as versions in substrate. For instance, the `StakingInterface` here seems ~90% the same as `sp_staking::StakingInterface`
- Is the functionality we need a subset of the functionality in `pallet-staking`? If so I think we should weigh the maintenance cost of copying/editing code vs just using `pallet-staking` and layering on top
